File: sdk_v2/js/src/openai/audioClient.ts
// `AudioClient`, layered on `AudioSession` via the OpenAI-JSON pass-through pattern. Mirrors C# `OpenAIAudioClient`.

import { Item, type Item as ItemT, type TextItem } from "../items.js";
import type { Model } from "../model.js";
import { Request } from "../request.js";
import type { Response } from "../response.js";
import { AudioSession } from "../session.js";
import { LiveAudioTranscriptionSession } from "./liveAudioSession.js";

// biome-ignore lint/suspicious/noExplicitAny: OpenAI request/response objects are user-shaped JSON.
type Json = any;

export class AudioClientSettings {
  language?: string;
  temperature?: number;

  /** @internal */
  _serialize(): Record<string, Json> {
    const result: Record<string, Json> = {};
    if (this.language !== undefined) result.language = this.language;
    if (this.temperature !== undefined) result.temperature = this.temperature;
    return result;
  }
}

export class AudioClient {
  readonly #model: Model;
  #disposed = false;

  public settings = new AudioClientSettings();

  /** @internal — construct via `model.createAudioClient()`. */
  constructor(model: Model) {
    this.#model = model;
  }

  get modelId(): string {
    return this.#model.id;
  }

  /** True once `dispose()` has been called on this client. */
  get disposed(): boolean {
    return this.#disposed;
  }

  /**
   * Mark this client as disposed. Idempotent. After disposal, `transcribe` / `transcribeStreaming` throw.
   * Sessions are scoped to a single request, so there is no long-lived native resource to release here.
   * `LiveAudioTranscriptionSession` instances created via `createLiveTranscriptionSession()` have their
   * own lifetime and must be disposed separately.
   */
  dispose(): void {
    this.#disposed = true;
  }

  [Symbol.dispose](): void {
    this.dispose();
  }

  #checkNotDisposed(): void {
    if (this.#disposed) {
      throw new Error("AudioClient: already disposed");
    }
  }

  /**
   * Transcribe an audio file. Path is forwarded to native and may be any
   * format the underlying model accepts.
   */
  async transcribe(filePath: string): Promise<Json> {
    this.#checkNotDisposed();
    if (typeof filePath !== "string" || filePath.trim() === "") {
      throw new Error("filePath must be a non-empty string.");
    }
    const requestJson: Record<string, Json> = {
      model: this.modelId,
      filename: filePath,
      ...this.settings._serialize(),
    };

    const request = new Request();
    request.addItem(Item.text(JSON.stringify(requestJson), "openai-json"));

    const session = new AudioSession(this.#model);
    let response: Response;
    try {
      response = await session.processRequest(request);
    } catch (err) {
      throw new Error(
        `Audio transcription failed for model '${this.modelId}': ${err instanceof Error ? err.message : String(err)}`,
        { cause: err },
      );
    } finally {
      session.dispose();
    }

    const text = findOpenAiJsonText(response.output);
    if (text === undefined) {
      throw new Error(
        `Audio transcription for model '${this.modelId}' returned no openai-json text item.`,
      );
    }
    return JSON.parse(text);
  }

  /** Streaming transcribe. Yields parsed JSON chunks as the model emits them. */
  transcribeStreaming(filePath: string): AsyncIterable<Json> {
    this.#checkNotDisposed();
    if (typeof filePath !== "string" || filePath.trim() === "") {
      throw new Error("filePath must be a non-empty string.");
    }
    const requestJson: Record<string, Json> = {
      model: this.modelId,
      filename: filePath,
      stream: true,
      ...this.settings._serialize(),
    };

    const model = this.#model;
    const modelId = this.modelId;
    return {
      async *[Symbol.asyncIterator](): AsyncIterator<Json> {
        const request = new Request();
        request.addItem(Item.text(JSON.stringify(requestJson), "openai-json"));

        const session = new AudioSession(model);
        try {
          for await (const item of session.processStreamingRequest(request)) {
            if (item.type !== "text") continue;
            const t = item as TextItem;
            if (t.textType !== "openai-json" || t.text === "") continue;
            yield JSON.parse(t.text);
          }
        } catch (err) {
          if (err instanceof Error && err.name === "AbortError") throw err;
          throw new Error(
            `Streaming audio transcription failed for model '${modelId}': ${err instanceof Error ? err.message : String(err)}`,
            { cause: err },
          );
        } finally {
          session.dispose();
        }
      },
    };
  }

  /**
   * Create a real-time audio transcription session for streaming PCM input.
   * Call `start()`, push PCM chunks via `append(chunk)`, consume parsed
   * transcription responses via `getStream()`, then `stop()` and `dispose()`.
   */
  createLiveTranscriptionSession(): LiveAudioTranscriptionSession {
    return new LiveAudioTranscriptionSession(this.#model);
  }
}

function findOpenAiJsonText(output: ReadonlyArray<ItemT>): string | undefined {
  for (const item of output) {
    if (item.type === "text" && (item as TextItem).textType === "openai-json") {
      return (item as TextItem).text;
    }
  }
  return undefined;
}

File: sdk_v2/cs/src/OpenAI/AudioClient.cs
// --------------------------------------------------------------------------------------------------------------------
// <copyright company="Microsoft">
//   Copyright (c) Microsoft. All rights reserved.
// </copyright>
// --------------------------------------------------------------------------------------------------------------------

namespace Microsoft.AI.Foundry.Local;

using System.Text.Json;

using Betalgo.Ranul.OpenAI.ObjectModels.ResponseModels;

using Microsoft.AI.Foundry.Local.Detail;
using Microsoft.AI.Foundry.Local.OpenAI;
using Microsoft.Extensions.Logging;

using NativeModel = Microsoft.AI.Foundry.Local.Detail.Native.Model;

/// <summary>
/// Audio Client that uses the OpenAI API.
/// Implemented using Betalgo.Ranul.OpenAI SDK types.
/// </summary>
[System.Obsolete(
    "OpenAIAudioClient is deprecated and will be removed at the end of 2026. " +
    "Use AudioSession instead. OpenAI types remain supported for the web-server path.",
    error: false)]
public class OpenAIAudioClient
{
    private readonly string _modelId;
    private readonly NativeModel _nativeModel;
    private readonly ILogger _logger;

    internal OpenAIAudioClient(string modelId, NativeModel nativeModel)
    {
        _modelId = modelId;
        _nativeModel = nativeModel;
        _logger = FoundryLocalManager.Instance.Logger;
    }

    /// <summary>
    /// Settings that are supported by Foundry Local
    /// </summary>
    public record AudioSettings
    {
        public string? Language { get; set; }
        public float? Temperature { get; set; }
    }

    /// <summary>
    /// Settings to use for audio transcription using this client.
    /// </summary>
    public AudioSettings Settings { get; } = new();

    /// <summary>
    /// Transcribe audio from a file.
    /// </summary>
    /// <param name="audioFilePath">
    /// Path to file containing audio recording.
    /// Supported formats: mp3
    /// </param>
    /// <param name="ct">Optional cancellation token.</param>
    /// <returns>Transcription response.</returns>
    public async Task<AudioCreateTranscriptionResponse> TranscribeAudioAsync(string audioFilePath,
                                                                             CancellationToken? ct = null)
    {
        return await Utils.CallWithExceptionHandlingAsync(() => TranscribeAudioImplAsync(audioFilePath, ct),
                                                          "Error during audio transcription.", _logger)
                                                         .ConfigureAwait(false);
    }

    /// <summary>
    /// Transcribe audio from a file with streamed output.
    /// </summary>
    /// <param name="audioFilePath">
    /// Path to file containing audio recording.
    /// Supported formats: mp3
    /// </param>
    /// <param name="ct">Cancellation token.</param>
    /// <returns>An asynchronous enumerable of transcription responses.</returns>
    public IAsyncEnumerable<AudioCreateTranscriptionResponse> TranscribeAudioStreamingAsync(
        string audioFilePath, CancellationToken ct)
    {
        return Utils.WrapStreamingExceptions(
            TranscribeAudioStreamingImplAsync(audioFilePath, ct),
            "Error during streaming audio transcription.", _logger, ct);
    }

    /// <summary>
    /// Create a real-time streaming transcription session.
    /// Audio data is pushed in as PCM chunks and transcription results are returned as an async stream.
    /// </summary>
    /// <returns>A streaming session that must be disposed when done.</returns>
    public LiveAudioTranscriptionSession CreateLiveTranscriptionSession()
    {
        return new LiveAudioTranscriptionSession(_modelId, _nativeModel);
    }

    private Task<AudioCreateTranscriptionResponse> TranscribeAudioImplAsync(string audioFilePath,
                                                                             CancellationToken? ct)
    {
        var requestJson = AudioTranscriptionCreateRequestExtended
            .FromUserInput(_modelId, audioFilePath, Settings)
            .ToJson();

        return NativeRequestRunner.RunAsync(
            _nativeModel,
            requestJson,
            json => JsonSerializer.Deserialize(json, JsonSerializationContext.Default.AudioCreateTranscriptionResponse)
                    ?? throw new FoundryLocalException("Failed to deserialize audio transcription response."),
            _logger,
            ct);
    }

    private IAsyncEnumerable<AudioCreateTranscriptionResponse> TranscribeAudioStreamingImplAsync(
        string audioFilePath, CancellationToken ct)
    {
        var requestJson = AudioTranscriptionCreateRequestExtended
            .FromUserInput(_modelId, audioFilePath, Settings)
            .ToJson();

        return NativeRequestRunner.RunStreamingAsync<AudioCreateTranscriptionResponse>(
            _nativeModel,
            requestJson,
            json => JsonSerializer.Deserialize(json, JsonSerializationContext.Default.AudioCreateTranscriptionResponse),
            _logger,
            "Error processing streaming audio transcription callback data.",
            "Error executing streaming audio transcription.",
            ct);
    }
}
